src/main/java/hudson/plugins/mercurial/MercurialSCMSource.java

```diff
@@ -204,6 +207,18 @@ public long lastModified() {
         return null;
     }
 
+    @NonNull
+    @Override
+    protected List<Action> retrieveActions(@NonNull SCMHead head,
+                                           @edu.umd.cs.findbugs.annotations.CheckForNull SCMHeadEvent event,
+                                           @NonNull TaskListener listener) throws IOException, InterruptedException {
+        // TODO for Mercurial 2.4+ check for the bookmark called @ and resolve that to determine the primary
+        if ("default".equals(head.getName())) {
+            return Collections.<Action>singletonList(new PrimaryInstanceMetadataAction());
```

Comment:
Please fix the API definition of `retrieveActions` to accept `List<? extends Action>` so that implementations need not use type qualifiers.